Valkey 8.1 blogpost (#227)
valkey-io/valkey-io.github.io: content/blog/2025-03-27-valkey-8-1-0-rc1.md

Reviewer: TODO: Rename this file when the publish date is decided.

Other sources of overhead in the TLS connection handling were identified, namely in the calls to `SSL_pending()` and `ERR_clear_error()` functions, which were being called in the main event thread. By offloading these functions to the I/O threads pool, a throughput improvement was achieved in some operations. For instance, it was observed a [10%](https://github.com/valkey-io/valkey/pull/1271) improvement in `SET` operations throughput, and a [22%](https://github.com/valkey-io/valkey/pull/1271) improvement in `GET` operations throughput.

Replication traffic efficiency was also improved in 8.1 by offloading the reading of replication stream on the replicas to the I/O thread pool which means they can serve more read traffic. On the primaries, replication stream writes are now offloaded to the I/O thread pool.

### Replication Improvements

Full syncs with TLS enabled benefit from an [18%](https://github.com/valkey-io/valkey/pull/1479 ) improvement by removing CRC checksummming.

The fork copy-on-write memory overhead is reduced by up to [47%](https://github.com/valkey-io/valkey/pull/905)
 

> ### Sorted set and hyperloglog and bitcount optimisations

Reviewer 2: ```suggestion
### Sorted set and hyperloglog and bitcount optimizations
```

> `ZRANK` command, which serves a popular usecase in operating Leaderboards, was optimised to perform up to [45%](https://github.com/valkey-io/valkey/pull/1389) faster, depending on the serted set size.

Reviewer 2: ```suggestion
`ZRANK` command, which serves a popular usecase in operating Leaderboards, was optimized to perform up to [45%](https://github.com/valkey-io/valkey/pull/1389) faster, depending on the sorted set size.
```

description= "Valkey 8.1 is now generally available! Come learn about the exciting improvements in performance, reliability, and observability that are available in this new version."
# 'authors' are the folks who wrote or contributed to the post.
# Each author corresponds to a biography file (more info later in this document)
authors= [ "rdias", "mvisser" ]
+++

The Valkey community is proud to unveil the first release candidate of Valkey 8.1,
a minor update designed to further enhance performance, reliability, and observability 
over Valkey 8.0 for all Valkey installations. 
In this blog, we'll dive a bit deeper into some of these areas 
and talk about the exciting features built for this release.

## Performance

There are many changes in 8.1 that improve Valkey's performance either by decreasing the latency of some operations, increasing the throughput of some operations, or reducing the memory footprint of the overall system.

One of the main changes responsible for several performance improvements is the new hashtable implementation that is used as the main key-value store in Valkey and also used in the Hash, Set, and Sorted Set data types implementation.

> ### The New Hashtable

Reviewer: Do you want to include this in the same blogpost?

I have a draft blog post just for this. I was going to finish it, but if you will include it here, then I don't need to finish it.

What's better, a separate blog post or combined? Regardless of which, we can help each other finish the content.

I was going to add the following diagram for the improved memory usage from https://github.com/valkey-io/valkey/issues/1835:

![memory usage graph](https://private-user-images.githubusercontent.com/20649182/426625839-ad094d9c-24fc-4bf9-8b44-48cd1abc20f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwODY4MjEsIm5iZiI6MTc0MzA4NjUyMSwicGF0aCI6Ii8yMDY0OTE4Mi80MjY2MjU4MzktYWQwOTRkOWMtMjRmYy00YmY5LThiNDQtNDhjZDFhYmMyMGYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI3VDE0NDIwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3OThkOTA3MWI1NTBkNWI3ODZhYWMzZjc4NmJmYjNmNjQ2YWRmMjQwNDM4NDRiNjE4OGI1OGQwNDE2ZjY1NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.N1qM6OGg1jBR-LCuR-gdIGB9YM5l7B9-2HSZPgk_oMY)

Reviewer 3: @zuiderkwast 

My plan was to talk about the new hashtable briefly in this blog post, since it's an highlight of the release, and then have your blog post dedicated just to the hashtable with all the details.

       | Metadata | Entry | Entry | Entry | Entry | Entry | Entry | Entry |
       +------------------------------------------------------------------+
      /            ` - - . _ _
     /                         `- - . _ _
    /                                     ` - . _
   +----------------------------------------------+
   | c ppppppp hash hash hash hash hash hash hash |
   +----------------------------------------------+
    |    |       |
    |    |      One byte of hash for each entry position in the bucket.
    |    |
    |   Presence bits. One bit for each entry position, indicating if an
    |   entry present or not.
    |
   Chained? One bit. If set, the last entry is a child bucket pointer.
```

The secondary hashes are used for quickly eliminating hash collisions when looking up a key. In this way, we can avoid comparing the key for a mismatching key, except once in 256.

This new design reduces by one the number of memory accesses required to lookup an entry when there are no collisions, but it reduces a lot more memory accesses to lookup an entry when there are key mapping collisions because each bucket can store 7 entries.

> It also reduces the memory footprint by a lot when there are collisions. With 3 keys mapping to the same bucket, the new design is already using less 8 bytes than the old design, which requires 72 bytes of memory (3 `dictEntry` objects). In the new design, the 64 byte bucket can store up to 7 entries, while the old design would require 168 bytes.

Reviewer 2: ```suggestion
It also reduces the memory footprint when there are collisions. With 3 keys mapping to the same bucket, the new design is already using less 8 bytes than the old design, which requires 72 bytes of memory (3 `dictEntry` objects). In the new design, the 64 byte bucket can store up to 7 entries, while the old design would require 168 bytes.
```
Avoid weasel words

Reviewer 4: I think that going into the bits of it is just bringing more confusion. suggesting to simplify by something like:

```
This new cache friendly design, requiers storing less memory pointers which reduces the memory footprint and the need to perform expensive memory dereferences.
```

> 

Reviewer 5: Should we mention Active Defrag performance improvements?  Maybe something like:

>Active Defrag has been improved to eliminate large latencies.  Latency impacts are now reduced below 1ms.  Active Defrag has also been improved in the presence of long-running commands with the introduction of anti-starvation protection.

And do we want to mention the change to the client cron?  Maybe:
>Periodic processing/monitoring of clients has been moved off of the 100ms serverCron eliminating the need for the `dynamic-hz` configuration.  This reduces overall system overhead when large numbers of clients are present.

> ### Other Performance Improvements

Reviewer 2: Do we need this other section? Feels like we could just leave everything under performance.

Several performance improvements were also added to the 8.1 release, here's a list of a few of them:

#### I/O threads improvements

Following up the I/O threads improvements added in 8.0, more operations have been offloaded to the I/O thread pool in the 8.1 release, improving the throughput and latency of some operations.

In the new 8.1 release, TLS connections are now able to efficiently use I/O threads for the several phases of the protocol. The TLS negotiation, which takes non-negliglibe amount of time to setup new connections, can now be offloaded to the I/O thread pool. This change has improved the rate of accepting new connections by around 300%.

Other sources of overhead in the TLS connection handling were identified, namely in the calls to `SSL_pending()` and `ERR_clear_error()` functions, which were being called in the main event thread. By offloading these functions to the I/O threads pool, a throughput improvement was achieved in some operations. For instance, it was observed a 10% improvement in `SET` operations throughput, and a 22% improvement in `GET` operations throughput.

> Replication traffic efficiency was also improved in 8.1 by offloading the reading of replication stream on the replicas to the I/O thread pool which means they can serve more read traffic. On the primaries, replication stream writes are now offloaded to the I/O thread pool.

Reviewer 4: I think replication had several improvements:
1. The same memory batching and prefetching techniques were applied to improve the BGSAVE operation time by 30%
2. We allow skipping the crc during replication which improve the save time by 18%
3. we also reduced the fork copy-on-write memory overhead by up to 47%

Author: @ranshid do you have the PR for item 1?

Reviewer 4: https://github.com/valkey-io/valkey/pull/1501 @martinrvisser , there you go

Reviewer: Yeah, that's iterator prefetching is what I was referring to in https://github.com/valkey-io/valkey-io.github.io/pull/227#issuecomment-2758895740. It's not specific to IO threading though. It appies to single-threaded as well.

#### Sorted set and hyperloglog optimisations 

> Valkey 8.1 continues to add optimisations which will greatly benefit users. Leaderboards are a popular use case for Valkey using its sorted set data type. Retrieving the specific rank of an entry on the leaderboard is done through the ZRANK operation. 
> 
> By analyzing the performance of ZRANK, a new algorithm was developed to optimise it’s performance by up to 45% depending on the size of the sorted set. 

Reviewer 4: ```suggestion
ZRANK command, which serves a popular usecase in operating Leaderboards, was optimised to perform up to 45% faster, depending on the serted set size.
```

calling it a new "algorithm" makes me feel uncomfortable.

Reviewer: It is a new algorithm though. It's some kind of reverse lookup. I remember we wanted to coin a term for this in the PR. I think it deserves a cool term. Let's call it *Ran's algortigm*?

> The probabilistic hyperloglog is another great data type, used for counting unique elements in very large datasets whilst using only 12KB of memory regardless of the amount of elements. By using the  modern CPUs Advanced Vector Extensions, Valkey 8.1 can achieve a 12x speed for the operations like PFMERGE and PFCOUNT on hyperloglog data types.

Reviewer 4: should we mention this was done for x86 arch only?

commandlog-large-reply-max-len : [ 128 ] number of latest entries to keep
```

Note : the existing SLOWLOG commands and parameters remain available. The slowlog parameters or the commandlog-execution parameters can be used interchangeably.

### Improved latency insights

Valkey has a built-in latency monitoring framework which samples latency-sensitive code paths such as for example fork when enabled through `latency-monitor-threshold` [latency monitor](https://valkey.io/topics/latency-monitor/). 

The new feature adds to additional metrics to the LATENCY LATEST command that reports on the latest latency events that have been collected. The additional information in Valkey 8.1 reports on the total of the recorded latencies as well as the number of recorded spikes for this event. These additional fields allow greater insight over Valkey 8 where only the maximum and latest spikes were recorded and reported.

## Extensability

Valkey is already well known by its extensability features. The sophisticated module system allows to extend the core system with new features developed as external modules.

> ### Programmability 

Reviewer 2: I think this work is very important, but there isn't very much a user can do on day one with it. My original thought it we should write a seperate blog about this work specifically in how it sets us up for the future and omit from 8.1. Not a strong opinion though.

Reviewer 3: We added this for the same reason as the new modules section. We wanted to highlight the feature but without giving to much detail because that should be relayed for a dedicated blog post.

But I can also cut this section if everyone agrees.

In Valkey 8.1 the module system API was extended with the support for developing new scripting engines as external modules.

This new API opens the door for the development of new language and runtime alternatives to the Lua base scripts supported by the Valkey core when using `EVAL` and `FCALL` commands.

In the future we expect the emergence of new scripting engines. A good candidate is a scripting engine based on WASM, allowing `EVAL` scripts to be written in other languages than Lua and to be executed in a more secure sandbox environment.

There are also benefits for existing Lua scripts, since new Lua runtimes can be easily plugged in that provide better security properties and/or better performance.

Developers that intend to build new scripting engines for Valkey should check the [Module API](https://valkey.io/topics/modules-api-ref/) documentation.

> ### New Modules

Reviewer 3: We decided to reference them just to create awareness about this. Briefly mentioning it here does not preclude having a dedicated blog post for the modules upon their release.

> ### New Modules

Reviewer 2: OK, but we need to clarify they are not in 8.1. They are a seperate release.

> ### New Modules

Reviewer 6: good point @madolson. I also feel that we should skip modules altogether in this blog.

> ### New Modules

Author: new modules removed

Frequently asked for use cases for Valkey are the JSON, Search and Bloom functionality. With Valkey 8.1 three new modules are available for early exploration. These modules will be bundled in an official release later this year, allowing developers to further customize their experience with Valkey.

- *Search* Module : a high-performance vector similarity search engine, contributed by Google, designed for AI-driven workloads demanding single-digit millisecond latency and high QPS. Capable of handling billions of vectors with demonstrated recall over 99%. Currently in beta with expected General Availability at the end of Q2.
- *JSON* Module : native support for the JSON data type, contributed by AWS, simplifying document-based storage and retrieval by allowing updating and retrieving a subset of the JSON data without having to process it on the client.
- Probabilistic *Bloom* Filter Module : efficient large-scale membership testing, contributed by AWS, reducing memory usage by over 90% for many applications including fraud detection and advertising.

> ## Additional Highlights

Author: We'll cut it down

+++

The Valkey community is proud to unveil the new release of Valkey 8.1,
a minor version update designed to further enhance performance, reliability, observability and usability
over Valkey 8.0 for all Valkey installations. 
In this blog, we'll dive a bit deeper into some of these areas 
and talk about the exciting features built for this release.

## Performance

There are many changes in 8.1 that improve Valkey's performance either by decreasing the latency of some operations, increasing the throughput of some operations and reducing the memory footprint of the overall system.

### The New Hashtable

One of the main changes responsible for several performance improvements is the new hashtable implementation that is used as the main key-value store in Valkey and also used in the Hash, Set, and Sorted Set data types implementation.

Valkey is a distributed key-value and therefore is no surprise that a hashtable data structure performs an important role in the way Valkey works.

Almost all commands go through the main key-value store where the several datatypes values are stored, and therefore the hashtable implementation is present in the critical path of most workloads.

> The new hashtable implementation is a complete rewrite of the previous hashtable implementation, which reduces the memory usage by roughly 20 bytes per key-value pair. The new implementation also helps improve the Valkey server supported throughput by roughly 10% compared to 8.0 version for pipeline workloads when I/O threading is not used.

Reviewer 6: > roughly 10% compared to 8.0 version for pipeline workloads when I/O threading is not used.

can we add a link to the issue or PR where 10% is reported, for more context?

Reviewer: I got this locally (laptop, loopback interface, benchmark and server on the same machine) but it wasn't consistently reproduced in proper benchmarks, so I didn't claim this in the bash table blog post. We should probably leave this out.

Btw, the hash table blog post is live so you can link to it.

The previous hashtable implementation, called `dict`, had the traditional design of an array of buckets where each bucket is a collision list. Each collision list element is called a `dictEntry` and holds a single key-value pair. The `dictEntry` structure stores three pointers, one for the key, one for the value, and one for the next `dictEntry` in the collision list. In terms of memory consumption, in a 64bit architecture, each `dictEntry` takes 24 bytes of memory.

![Dictionary structure](/assets/media/pictures/dict_structures.png)

Regarding memory accesses, the above design requires at least 4 memory accesses to get the value in the case that the key is stored in the first `dictEntry` of the collision list. This might have a significant impact in cache locality when traversing the collision list search for the right key.

> The new hashtable implementation improves on the memory footprint required and on the cache locality by taking a different approach on the design of each bucket. Each bucket has a fixed size of 64 bytes (in a 64bit architecture), which usually corresponds to the size of a single cache line. The bucket structure stores 7 pointers, 8 bytes each, plus 8 bytes used for metadata. The 7 pointers are used to store up to 7 entries. Keys that map to the same bucket are all stored in the same bucket. If more than 7 keys map to the same bucket, a new bucket is linked to the previous one using the last of the 7 pointers. Each entry pointer, points to an object that may embed the key and the value.

Reviewer 6: there is quite some detail in these paragraphs. @zuiderkwast, how do you like to cover this part to reduce redundancy? I was thinking that we could focus more on the results/impacts and the high level ideas in this blog and leave the details like data structures etc to yours. this would also align better with other sections such as "Iterator Prefetching"

Reviewer: Yeah I think we should remove most of the text about the internals of the hash table. Only the key takeaway are needed in this release blog post and a link to the hash table dedicated blog post.

The fork copy-on-write memory overhead is reduced by up to [47%](https://github.com/valkey-io/valkey/pull/905)
 
### Sorted set and hyperloglog and bitcount optimisations

`ZRANK` command, which serves a popular usecase in operating Leaderboards, was optimised to perform up to [45%](https://github.com/valkey-io/valkey/pull/1389) faster, depending on the serted set size.

The probabilistic hyperloglog is another great data type, used for counting unique elements in very large datasets whilst using only 12KB of memory regardless of the amount of elements. By using the modern CPUs Advanced Vector Extensions of x86, Valkey 8.1 can achieve a [12x](https://github.com/valkey-io/valkey/pull/1293) speed for the operations like `PFMERGE` and `PFCOUNT` on hyperloglog data types.

Similarly, the BITCOUNT operation has been improved up to [514%](https://github.com/valkey-io/valkey/pull/1741) using AVX2 on x86.

### Other Performance Enhancements

Active Defrag has been improved to eliminate large latencies to below [1ms](https://github.com/valkey-io/valkey/pull/1242) and anti-starvation protection has been introduced in the presence of long-running commands.

> Periodic processing/monitoring of clients has been moved off of the 100ms `serverCron()` interval loop, reducing system [overhead](https://github.com/valkey-io/valkey/pull/1387)
>  when large numbers of clients are present. 

Reviewer 2: ```suggestion
```
I don't think people will follow this.

Author: rewritten

## Observability

> There are also several improvements to the observability of the system behavior in Valkey 8.1.

Reviewer 6: Maybe expand a bit in the observability area to name a few new metrics added and a short description why they matter?

### Log Improvements

Valkey 8.1 brings new options to the format of the log file entries as well as the way timestamps are recorded in the log file. This makes it easier to consume the log files by log collecting systems. 

The format of the log file entries is controlled by the `log-format` parameter, where the default is the existing format :

- `legacy`: the default, traditional log format
- `logfmt`: a structured log format; see https://www.brandur.org/logfmt

The formatting of the timestamp of the log file entries is controlled by the `log-timestamp-format` parameter, where the default is the existing format: 

- `legacy`: default format
- `iso8601`: ISO 8601 extended date and time with time zone, of the form yyyy-mm-ddThh:mm:ss.sss±hh:mm
- `milliseconds`: milliseconds since the epoch

> An example of the format `logfmt` with the timestamp format `iso8601`:

Reviewer 6: I like everything til the examples - I would suggest omitting these examples (either to their own blog or links to the PR/issues).

```
pid=3883950 role=primary timestamp="2025-03-26T05:17:30.402-04:00" level=notice message="oO0OoO0OoO0Oo Valkey is starting oO0OoO0OoO0Oo
```

An example of the `legacy` format which remains the same:

```
3223494:M 18 Mar 2025 14:47:42.070 * oO0OoO0OoO0Oo Valkey is starting oO0OoO0OoO0Oo:
```

*Note*: as can be observed from the examples, using both the `logfmt` and `iso8601` format uses around 60% more space, so disk space should be considered when implementing these.

### Extending the Slowlog to Commandlog

Valkey has long had the capability to record slow commands during execution time based on the threshold set with  `slowlog-log-slower-than` parameter to keep the last `slowlog-max-len` entries.

The new `COMMANDLOG` feature expands the recording of slow commands beyond execution time by adding recording of large requests and replies. These metrics do not necessarily affect execution time but do affect the overall round trip to the application and impact network usage.

> The following parameters configure the recording of `COMMANDLOG`:

Reviewer 6: similar to my comments on the log format, I think the key message of  a release blog like this is to excite the audience. The "instructions" or "manual" are important too but can dilute the key message.

## Extensability

Valkey is already well known by its extensability features. The sophisticated module system allows to extend the core system with new features developed as external modules.

### Programmability 

In Valkey 8.1 the module system API was extended with the support for developing new scripting engines as external modules.

This new API opens the door for the development of new language and runtime alternatives to the Lua base scripts supported by the Valkey core when using `EVAL` and `FCALL` commands.

In the future we expect the emergence of new scripting engines. A good candidate is a scripting engine based on WASM, allowing `EVAL` scripts to be written in other languages than Lua and to be executed in a more secure sandbox environment.

There are also benefits for existing Lua scripts, since new Lua runtimes can be easily plugged in that provide better security properties and/or better performance.

Developers that intend to build new scripting engines for Valkey should check the [Module API](https://valkey.io/topics/modules-api-ref/) documentation.

### New Modules

> Frequently asked for use cases for Valkey are the JSON, Search and Bloom functionality. With Valkey 8.1 three new modules are available for early exploration. These modules will be bundled in an official release later this year, allowing developers to further customize their experience with Valkey.

Reviewer 2: Need to clarify these are not natively part of 8.1, but a secondary release. I actually feel more strongly we should just omit these. We have other vehicles to raise awareness about them. I really don't want someone to get confused about what is in the core release.

Reviewer 6: +1

Author: Removing

- **Search Module**: a high-performance vector similarity search engine, contributed by Google, designed for AI-driven workloads demanding single-digit millisecond latency and high QPS. Capable of handling billions of vectors with demonstrated recall over 99%. Currently in beta with expected General Availability at the end of Q2.
- **JSON Module**: native support for the JSON data type, contributed by AWS, simplifying document-based storage and retrieval by allowing updating and retrieving a subset of the JSON data without having to process it on the client.
- **Probabilistic Bloom Filter Module**: efficient large-scale membership testing, contributed by AWS, reducing memory usage by over 90% for many applications including fraud detection and advertising.

## Additional Highlights

> ### `valkey-benchmark` Reading From Replicas Option
> 
> `valkey-benchmark` is a commonly used tool to baseline Redis and Valkey performance to compare between different software versions, different hardware or deployments etc. It already had a cluster-aware mode but did not have the ability to run against any replicas as part of the benchmark. Valkey replicas are deployed in production to increase availability and can serve read operations. With the Valkey 8.1 addition to read from replicas to valkey-benchmark, this allows for benchmark against these more production like deployments.

Reviewer 2: ```suggestion
```
You gave a +1 to the previous comment but it's still here, do you want to keep it?

Author: I cut it down previously. Compared to the other improvements it is minor but worth highlighting as it has a decent value for users?

Reviewer 2: If you are an end user of Valkey, you will likely have *never* used the benchmark. It's used far more by developers and contributors for small targeted it. I think the intended audience is end users, so want to try to maximize the amount of things they will find interesting.

### Conditional Updates

This new functionality allows Valkey users to perform conditional updates using the `SET` command if the given comparison-value matches the key’s current value. This is a not only a quality-of-life improvement for developers as they no longer need to add this condition to their application code, it also saves a roundtrip to first get a value and then compare it before a `SET`. When using the optional `GET` as part of the `SET IFEQ`, the existing value is returned regardless whether it matches the comparison-value.

Syntax:
```
SET key value [ NX | XX | IFEQ comparison-value ] [ GET ] [ EX seconds | PX milliseconds | EXAT unix-time-seconds | PXAT unix-time-milliseconds | KEEPTTL ]

> ```

Reviewer 6: I would suggest skipping the tactical information.
